### PR TITLE
[rozhlas] Support the site update

### PR DESCRIPTION
**Pull request by amra**
Rozhlas sites changed way of playing audio on their sites. Supported sites are:
- vltava.rozhlas.cz
- wave.rozhlas.cz
- radiozurnal.rozhlas.cz
- dvojka.rozhlas.cz
- plus.rozhlas.cz
- sport.rozhlas.cz
- d-dur.rozhlas.cz
- jazz.rozhlas.cz
- junior.rozhlas.cz
- pohoda.rozhlas.cz
- english.radio.cz

Leaving the old player in place, because it's working old links.

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- ~~[ ] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests~~
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- ~~[ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)~~

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- ~~[ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))~~
- ~~[ ] Core bug fix/improvement~~
- ~~[ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))~~

### CODE REVIEW

**Comment by pukkandan** [comment]
> * [x]  At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2023-01-04T18%3A09%3A38Z#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2023-01-04T18%3A09%3A38Z#yt-dlp-coding-conventions)

Pls do so again

**Comment by amra** [comment]
> > * [x]   At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2023-01-04T18%3A09%3A38Z#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2023-01-04T18%3A09%3A38Z#yt-dlp-coding-conventions)
> 
> Pls do so again

Code improved. Please check again.

**Question**: I prefer to use a different id then created by `self._match_id(url)`, but then test wouldn't pass. Is there a way to use a dedicated id?
I prefer to use entry.get('meta').get('ga').get('contentId').

File: yt_dlp/extractor/rozhlas.py [hunk 1]
```diff
@@ -45,3 +47,70 @@ def _real_extract(self, url):
...
+
+class RozhlasVltavaIE(InfoExtractor):
+    _VALID_URL = r'https?://((vltava|wave|radiozurnal|dvojka|plus|sport|d-dur|jazz|junior|pohoda)\.rozhlas|english\.radio)\.cz/[a-z|-]*-(?P<id>\d+)'
+    _TESTS = [{
+        'url': 'https://wave.rozhlas.cz/papej-masicko-porcujeme-a-bilancujeme-filmy-a-serialy-ktere-letos-zabily-8891337',
+        'md5': 'ba2fdbc1242fc16771c7695d271ec355',
+        'info_dict': {
+            'id': '8891337',
+            'ext': 'mp3',
+            'title': 'Máte po celé sezóně poslouchání Čelistí právo na wellness, nebo můžete být rádi, že jste vůbec naživu? Poslechněte si filmově kritickou bilanci všeho, co v uplynulých měsících stálo za vidění',
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:60]:
```py
            'title': 'md5:xxx',
```

File: yt_dlp/extractor/rozhlas.py [hunk 2]
```diff
@@ -45,3 +47,70 @@ def _real_extract(self, url):
...
+                playerDiv = line.strip()
+                break
+
+        if playerDiv.count('<div') > 1:
+            for k, element in enumerate(playerDiv.split('<div')):
+                if element.count('mujRozhlasPlayer') == 1:
+                    playerDiv = '<div' + element
+                    break
+
+        return playerDiv
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:81]: use regex if get_element_text_and_html_by_tag doesnt work

File: yt_dlp/extractor/rozhlas.py [hunk 3]
```diff
@@ -45,3 +47,70 @@ def _real_extract(self, url):
...
+
+    def _real_extract(self, url):
+        webpage = self._download_webpage(url, None)
+
+        playerDiv = self.find_element(webpage)
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:86]: snake_case everywhere

File: yt_dlp/extractor/rozhlas.py [hunk 4]
```diff
@@ -45,3 +47,70 @@ def _real_extract(self, url):
...
+
+        jsonString = extract_attributes(playerDiv).get('data-player')
+        jsonData = json.loads(jsonString)
+
+        entries = []
+        for entry in jsonData.get("data").get('playlist'):
+            entry0 = entry.get('audioLinks')[0]
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:93]: why only 0?

File: yt_dlp/extractor/rozhlas.py [hunk 5]
```diff
@@ -45,3 +47,70 @@ def _real_extract(self, url):
...
+                'vcodec': 'none',
+            }
+            temp = {
+                'id': self._match_id(url),  # Prefering to user entry.get('meta').get('ga').get('contentId') for id. Using this because of tests.
+                'title': entry.get('title'),
+                'description': entry.get('meta').get('ga').get('contentEvent'),
+                'duration': entry.get('duration'),
+                'formats': [format],
+            }
+            entries.append(temp)
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:109]:
```py
            entries.append({
                ...
                'formats': [{
                    ...
                }],
            })
```

File: yt_dlp/extractor/rozhlas.py [hunk 6]
```diff
@@ -45,3 +47,70 @@ def _real_extract(self, url):
...
+
+        return {
+            '_type': 'playlist',
+            'id': jsonData.get('data').get('embedId'),
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:113]: https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#mandatory-and-optional-metafields

File: yt_dlp/extractor/rozhlas.py [hunk 7]
```diff
@@ -45,3 +47,70 @@ def _real_extract(self, url):
...
+            format = {
+                'url': entry0.get('url'),
+                'ext': entry0.get('variant'),
+                'format_id': entry0.get('variant'),
+                'abr': entry0.get('bitrate'),
+                'acodec': entry0.get('variant'),
+                'vcodec': 'none',
+            }
+            temp = {
+                'id': self._match_id(url),  # Prefering to user entry.get('meta').get('ga').get('contentId') for id. Using this because of tests.
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:103]: pls explain

File: yt_dlp/extractor/rozhlas.py [hunk 8]
```diff
@@ -45,3 +47,70 @@ def _real_extract(self, url):
...
+        if playerDiv.count('<div') > 1:
+            for k, element in enumerate(playerDiv.split('<div')):
+                if element.count('mujRozhlasPlayer') == 1:
+                    playerDiv = '<div' + element
+                    break
+
+        return playerDiv
+
+    def _real_extract(self, url):
+        webpage = self._download_webpage(url, None)
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:84]:
```suggestion
        video_id = self._match_id(url)
        webpage = self._download_webpage(url, video_id)
```

File: yt_dlp/extractor/rozhlas.py (hunk 3 above)
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:86]: Does this need to be a function?

File: yt_dlp/extractor/rozhlas.py [hunk 9]
```diff
@@ -45,3 +47,70 @@ def _real_extract(self, url):
...
+
+        playerDiv = self.find_element(webpage)
+
+        jsonString = extract_attributes(playerDiv).get('data-player')
+        jsonData = json.loads(jsonString)
+
+        entries = []
+        for entry in jsonData.get("data").get('playlist'):
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:92]:
```suggestion
        data = self._parse_json(
            extract_attributes(playerDiv)['data-player'], video_id)['data']

        entries = []
        for entry in data['playlist']:
```

File: yt_dlp/extractor/rozhlas.py [hunk 10]
```diff
@@ -45,3 +47,70 @@ def _real_extract(self, url):
             'duration': duration,
             'vcodec': 'none',
         }
+
+
+class RozhlasVltavaIE(InfoExtractor):
+    _VALID_URL = r'https?://((vltava|wave|radiozurnal|dvojka|plus|sport|d-dur|jazz|junior|pohoda)\.rozhlas|english\.radio)\.cz/[a-z|-]*-(?P<id>\d+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:53]: https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#keep-the-regular-expressions-as-simple-as-possible-but-no-simpler

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
> **Question**: I prefer to use a different id then created by `self._match_id(url)`, but then test wouldn't pass. Is there a way to use a dedicated id?

It will, just set the same id in test definition

File: yt_dlp/extractor/rozhlas.py (hunk 4 above)
Comment (amra) [reply yt_dlp/extractor/rozhlas.py:93]: They support just one format (mp3). I didn't find anything I can test it on.

File: yt_dlp/extractor/rozhlas.py (hunk 1 above)
Comment (amra) [reply yt_dlp/extractor/rozhlas.py:60]: done

File: yt_dlp/extractor/rozhlas.py (hunk 3 above)
Comment (amra) [reply yt_dlp/extractor/rozhlas.py:86]: done

File: yt_dlp/extractor/rozhlas.py (hunk 5 above)
Comment (amra) [reply yt_dlp/extractor/rozhlas.py:109]: done

File: yt_dlp/extractor/rozhlas.py (hunk 8 above)
Comment (amra) [reply yt_dlp/extractor/rozhlas.py:84]: done

File: yt_dlp/extractor/rozhlas.py (hunk 2 above)
Comment (amra) [reply yt_dlp/extractor/rozhlas.py:81]: done

File: yt_dlp/extractor/rozhlas.py (hunk 3 above)
Comment (amra) [reply yt_dlp/extractor/rozhlas.py:86]: done

File: yt_dlp/extractor/rozhlas.py (hunk 6 above)
Comment (amra) [reply yt_dlp/extractor/rozhlas.py:113]: done

It's set to `None` if there are multiple entries.

File: yt_dlp/extractor/rozhlas.py (hunk 10 above)
Comment (amra) [reply yt_dlp/extractor/rozhlas.py:53]: done

**Comment by amra** [comment]
> It will, just set the same id in test definition

I mean the `id` in `entries` list.

File: yt_dlp/extractor/rozhlas.py [hunk 11]
```diff
@@ -45,3 +47,54 @@ def _real_extract(self, url):
...
+        'url': 'https://wave.rozhlas.cz/papej-masicko-porcujeme-a-bilancujeme-filmy-a-serialy-ktere-letos-zabily-8891337',
+        'md5': 'ba2fdbc1242fc16771c7695d271ec355',
+        'info_dict': {
+            'id': '8891337',
+            'ext': 'mp3',
+            'title': 'md5:1c6d29fb9564e1f17fc1bb83ae7da0bc',
+            'description': 'md5:bbab1431c3472304b4863faea414cd2b',
+            'duration': 1574
+        }
+    }]
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:64]: Definition is wrong. This happens to pass the test due to a quirk (bug?) in the test runner. Add `playlist` key to the definition. E.g. https://github.com/yt-dlp/yt-dlp/blob/ce03e9b90acef7a8b59b1193e32cbd9898c93ad0/yt_dlp/extractor/afreecatv.py#L51-L79

Comment (pukkandan) [reply yt_dlp/extractor/rozhlas.py:64]: This should also fix what you are saying about id

Comment (amra) [reply yt_dlp/extractor/rozhlas.py]: done

The id problem was a bug on my side. 

I added a test for series download.

File: yt_dlp/extractor/rozhlas.py (hunk 6 above)
Comment (pukkandan) [reply yt_dlp/extractor/rozhlas.py]: Unaddressed. You realize `.get(...).get(...)` is useless to make something non-fatal, right?

image
Refer above link for details and examples

Comment (amra) [reply yt_dlp/extractor/rozhlas.py]: fixed

File: yt_dlp/extractor/rozhlas.py (hunk 7 above)
Comment (amra) [reply yt_dlp/extractor/rozhlas.py]: removed

File: yt_dlp/extractor/rozhlas.py [hunk 12]
```diff
@@ -45,3 +48,142 @@ def _real_extract(self, url):
...
+                'chapter_number': int(traverse_obj(entry, ('meta', 'ga', 'contentSerialPart'))) if traverse_obj(entry, ('meta', 'ga', 'contentSerialPart')) is not None else None,
+            })
+
+        return {
+            '_type': 'playlist',
+            'id': data.get('embedId'),
+            'title': traverse_obj(data, ('series', 'title')),
+            'formats': entries[0].get('formats') if len(entries) == 1 else None,
+            'entries': entries,
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:189]:
```suggestion

        return {
            '_type': 'playlist',
            'id': data.get('embedId'),
            'title': traverse_obj(data, ('series', 'title')),
            'entries': [{
                'id': entry['meta']['ga']['contentId'],
                'title': traverse_obj(entry, ('meta', 'ga', 'contentName')),
                'description': entry.get('title'),
                'duration': entry.get('duration'),
                'artist': traverse_obj(entry, ('meta', 'ga', 'contentAuthor')),
                'channel_id': traverse_obj(entry, ('meta', 'ga', 'contentCreator')),
                'chapter': traverse_obj(entry, ('meta', 'ga', 'contentNameShort')),
                'chapter_number': int_or_none(traverse_obj(entry, ('meta', 'ga', 'contentSerialPart'))),
                'formats': [{
                    'url': audio_link['url'],
                    'ext': audio_link.get('variant'),
                    'format_id': audio_link.get('variant'),
                    'abr': audio_link.get('bitrate'),
                    'acodec': audio_link.get('variant'),
                    'vcodec': 'none',
                } for audio_link in entry['audioLinks']],
            } for entry in data['playlist']],
        }
```

File: yt_dlp/extractor/rozhlas.py [hunk 13]
```diff
@@ -45,3 +48,142 @@ def _real_extract(self, url):
...
+            'id': 8891337,
+            'title': 'md5:21f99739d04ab49d8c189ec711eef4ec',
+        },
+        'playlist_count': 1,
+        'playlist': [{
+            'md5': 'ba2fdbc1242fc16771c7695d271ec355',
+            'info_dict': {
+                'id': '10520988',
+                'ext': 'mp3',
+                'title': "Papej masíčko! Porcujeme a bilancujeme filmy a seriály, které to letos zabily",
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:68]:
```suggestion
                'title': 'Papej masíčko! Porcujeme a bilancujeme filmy a seriály, které to letos zabily',
```

File: yt_dlp/extractor/rozhlas.py [hunk 14]
```diff
@@ -45,3 +48,142 @@ def _real_extract(self, url):
...
+                'duration': 1574,
+                'artist': 'Aleš Stuchlý',
+                'channel_id': 'radio-wave',
+            },
+        }]
+    }, {
+        'url': 'https://wave.rozhlas.cz/poslechnete-si-neklid-podcastovy-thriller-o-vine-strachu-a-vztahu-ktery-zasel-8554744',
+        'info_dict': {
+            'id': 8554744,
+            'title': "Poslechněte si Neklid. Podcastový thriller o vině, strachu a vztahu, který zašel příliš daleko",
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:79]:
```suggestion
            'title': 'Poslechněte si Neklid. Podcastový thriller o vině, strachu a vztahu, který zašel příliš daleko',
```

**Comment by pukkandan** [comment]
test the changes and let me know

File: yt_dlp/extractor/rozhlas.py [hunk 15]
```diff
@@ -45,3 +48,135 @@ def _real_extract(self, url):
...
+            },
+        }]
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        # FIXME: Use get_element_text_and_html_by_tag when it accepts less strict html
+        player_div = self._parse_json(extract_attributes(self._search_regex(
```
Comment (pukkandan) [inline yt_dlp/extractor/rozhlas.py:155]:
```suggestion
        data = self._parse_json(extract_attributes(self._search_regex(
```

**Comment by amra** [comment]
I couldn't find why `_search_regex()` is not working. I switched back to `re.findall()`.

**Pull request merged**